### ISSUE TITLE
RingBufferWorkProcessor with EventBus implematation

### ISSUE THREAD

**Issue by aviadhadida** [reactor/reactor#495]
Hi.
I am trying to achieve something that i am not sure if this can be done.. I had a several converstaion with Stepahne and Jon about that issue. After reading the doc and trying a bit i feel that what i am doing is really bad practice. Does this thing is valid:

```
private <T> void subscribeToEvent(Selector<String> selector, Consumer<Event<T>> handler) {
    Processor<Event<T>, Event<T>> p = RingBufferWorkProcessor.create("test", 32);
    Stream<Event<T>> stream = Streams.wrap(p);
    for (int i = 0; i < 3; i++) {
        stream.consume(handler);
    }
    eventBus.on(selector, new Consumer<Event<T>>() {

        @Override
        public void accept(Event<T> t) {
            p.onNext(t);
        }
    });
}
```

As a reminder i am implementing(Refactoring) real time web application that client is big banks for trading(Forex,Bonds, Repo etc..)

What i have done here it to change subscription to event bus by adding work processor(In this example is 3 for every subscription).
I am starting to understand the Streams more properly and my feel is that i need to use event bus for some actions, or create several event buses for every module or something like that(This is hard to maintain). My problem is that event bus is single threaded and cannot be used for **blocking actions**.. 
The other thing i can do is to used the processor in order to handle real time data. For example get the stream from api. **Put on processor -> send to place to add fee-> send to place of user connection holder and write to channel**

So basically i need this for processing real time with a lot of data, maybe get from stream every 250ms only one message for each currency pair(EUR/USD, USD/ILS etc..)

Maybe can you help me.. I am really strugling here and cannot find the right way to implement this and how to use all of this framework in the best way.

Thanks a lot
Aviad

**Comment by aviadhadida** [comment]
I also tried to send my questions in stackoverflow.. If you can take a look please. Thanks
http://stackoverflow.com/questions/30980528/spring-reactor-implement-streaming-system-best-practice

**Comment by smaldini** [comment]
I'll try to find time, its end of Q2 and I have a bunch of tasks on the plate which is going to clear up next week.
TL'DR is that your understanding looks correct and it does totally make sense in your use case. There are ways to optimize your flow in details I'll elaborate on but the base is solid IMO.

**Comment by aviadhadida** [comment]
Thanks. Mean while i am keep on reading hopefully i will find the way of understanding the framework better.. Please update me when ever you can.. I really need help here :) Thanks!

**Comment by aviadhadida** [comment]
@smaldini Maybe you can help me here.. I was patient  :). I really need help here. We are making progress on our development and are starting to get stuck.. Please help..
Thanks a lot

**Comment by smaldini** [comment]
Sorry for the delay, we're limited resources wise especially I have customer engagements in first priority before anything, a situation that might evolve later on.

It's quite hard to establish what we can help with tbh at this level without talking face to face and using a nice whiteboard and take a dive in the architecture.

So for the basics: 
1-level  non blocking , no composition, dynamic routing : eventBus
N-levels non blocking, composition, typed routing : stream. Although using stream.groupBy you can usually replicate the behavior of a selector/key. 

Reactor 2.1 will give more love to the event bus which hasn't been yet reviewed since Reactor 1.

Now in terms of designs you don't want to create a Processor for every single pair, it usually matches up with physical constraints : 1 keep-alive connection -> processor, rather than 1-type of message out of N -> 1 Processor. Creating threads on the JVM doesn't scale. You could share the Executor between the various processor using the right create(ExecutorService) factory method but that only applies if the processor frees the thread at some point (onComplete or cancel from downstream).
In the end you really want to scale up the work so it doesn't matter that you create a unique processor per event key. I would give a try to invert the flow by putting WorkProcessor first, subscribing to it N times you want to parallelize, then notify the event Bus. This event bus doesn't need to be async anymore, just use a synchronous dispatcher on him, you need it mainly for routing. Then inside each bus consumer you can use a synchronous stream as well if you need more composition (Broadcaster.create(), which will be rebranded in 2.1 SyncProcessor so you get the idea).

**Comment by aviadhadida** [comment]
Thanks for responding.. 
I have decided to just use ExecutorService on the consumers and this way i can have consumer that can insert into DB or do anything else without blocking other consumers..

Really thanks for your help.. I will keep watching on this great framework and probably use Streams in my next phase.

Thanks a lot
